### PR TITLE
[MM-16709] Add helper method for plugins using MessageHasBeenPosted 

### PR DESCRIPTION
**Pull request by RajatVaryani**
#### Summary
The PR introduces a method in helper interface. The method has common code for filtering to be used across plugins which use `MessageHasBeenPosted`.

#### Ticket Link
Fixes https://github.com/mattermost/mattermost-server/issues/12392

### CODE REVIEW

File: plugin/helpers_bots.go [hunk 1]
```diff
@@ -72,3 +113,67 @@ func (p *HelpersImpl) EnsureBot(bot *model.Bot) (retBotId string, retErr error)
 
 	return createdBot.UserId, nil
 }
+
+func (p *HelpersImpl) ShouldProcessMessage(post *model.Post, botUserId string, options ...ShouldProcessMessageOption) bool {
```
Comment (hanzei) [inline plugin/helpers_bots.go]: I don't like the required `botUserId` parameter. There are cases where you don't need to provide it, like just wanting to skip all system messages. What do you think about adding `botUserId` as a parameter to `AllowSelf()`?

File: plugin/helpers_bots.go [hunk 2]
```diff
@@ -72,3 +113,67 @@ func (p *HelpersImpl) EnsureBot(bot *model.Bot) (retBotId string, retErr error)
...
+	if !messageProcessOptions.AllowSystemMessages {
+		if post.IsSystemMessage() {
+			return false
+		}
+	}
+
+	if !contains(messageProcessOptions.FilterChannelIDs, post.ChannelId) {
+		channel, _ := p.API.GetChannel(post.ChannelId)
+
+		if !IsBotDMChannel(channel, botUserId) {
```
Comment (hanzei) [inline plugin/helpers_bots.go]: This seams like a special case to me. Currently you skip every message that is not a DM to the bot by default. That's a  bit to restrictive. Maybe add a new `LimitToDM(userID string)` method? Or just rely on the plugin using `messageProcessOptions.FilterChannelIDs`?

File: plugin/helpers_bots.go [hunk 3]
```diff
@@ -72,3 +113,67 @@ func (p *HelpersImpl) EnsureBot(bot *model.Bot) (retBotId string, retErr error)
...
+	}
+
+	if !messageProcessOptions.AllowSystemMessages {
+		if post.IsSystemMessage() {
+			return false
+		}
+	}
+
+	if !contains(messageProcessOptions.FilterChannelIDs, post.ChannelId) {
+		channel, _ := p.API.GetChannel(post.ChannelId)
```
Comment (hanzei) [inline plugin/helpers_bots.go]: The error should not be ignored. Please [`Wrap`](https://godoc.org/github.com/pkg/errors#Wrap) it and return it instead.

File: plugin/helpers_bots.go [hunk 4]
```diff
@@ -72,3 +113,67 @@ func (p *HelpersImpl) EnsureBot(bot *model.Bot) (retBotId string, retErr error)
...
+			if user.IsBot {
+				return false
+			}
+		}
+	}
+
+	return true
+}
+
+func initialiseOptions(options []ShouldProcessMessageOption) *shouldProcessMessageOptions {
```
Comment (hanzei) [inline plugin/helpers_bots.go]: Nit: This can be part of `ShouldProcessMessage`.

File: plugin/helpers_bots.go [hunk 5]
```diff
@@ -72,3 +113,67 @@ func (p *HelpersImpl) EnsureBot(bot *model.Bot) (retBotId string, retErr error)
...
+	}
+
+	if !strings.HasPrefix(channel.Name, botUserID+"__") && !strings.HasSuffix(channel.Name, "__"+botUserID) {
+		return false
+	}
+
+	return true
+}
+
+func contains(a []string, x string) bool {
```
Comment (hanzei) [inline plugin/helpers_bots.go]: You can use `utils.StringInSlice` instead

File: plugin/helpers_bots.go [hunk 6]
```diff
@@ -72,3 +113,67 @@ func (p *HelpersImpl) EnsureBot(bot *model.Bot) (retBotId string, retErr error)
...
+
+func initialiseOptions(options []ShouldProcessMessageOption) *shouldProcessMessageOptions {
+	messageProcessOptions := &shouldProcessMessageOptions{}
+	for _, option := range options {
+		option(messageProcessOptions)
+	}
+	return messageProcessOptions
+}
+
+func IsBotDMChannel(channel *model.Channel, botUserID string) bool {
```
Comment (hanzei) [inline plugin/helpers_bots.go]: I think this should be part of `model`. Would you please move it there?

File: plugin/helpers.go [hunk 7]
```diff
@@ -37,6 +37,8 @@ type Helpers interface {
 	//
 	// Minimum server version: 5.6
 	KVSetWithExpiryJSON(key string, value interface{}, expireInSeconds int64) error
+
+	ShouldProcessMessage(post *model.Post, botUserId string, options ...ShouldProcessMessageOption) bool
```
[the remark this answers is not in the document]
Comment (RajatVaryani) [reply plugin/helpers.go]: Done.

File: plugin/helpers_bots.go (hunk 1 above)
Comment (RajatVaryani) [reply plugin/helpers_bots.go]: Point taken. Just trying to understand. The only problem occurs when I don't pass `AllowSelf()` then the `botUserID` won't be available for comparison.

File: plugin/helpers_bots.go (hunk 3 above)
Comment (RajatVaryani) [reply plugin/helpers_bots.go]: Done

File: plugin/helpers_bots.go (hunk 4 above)
Comment (RajatVaryani) [reply plugin/helpers_bots.go]: Agree

File: plugin/helpers_bots.go (hunk 5 above)
Comment (RajatVaryani) [reply plugin/helpers_bots.go]: Done

File: plugin/helpers_bots.go (hunk 6 above)
Comment (RajatVaryani) [reply plugin/helpers_bots.go]: Done

**Comment by RajatVaryani** [comment]
@hanzei Thanks for the feedback. Implemented some of the changes.

File: plugin/helpers_bots.go (hunk 1 above)
Comment (hanzei) [reply plugin/helpers_bots.go]: You only need `botUserID` for `AllowSelf` right? See also by comment https://github.com/mattermost/mattermost-server/pull/12401#discussion_r329481415

File: plugin/helpers_bots.go [hunk 8]
```diff
@@ -72,3 +112,52 @@ func (p *HelpersImpl) EnsureBot(bot *model.Bot) (retBotId string, retErr error)
...
+		if post.IsSystemMessage() {
+			return false, nil
+		}
+	}
+
+	if !utils.StringInSlice(post.ChannelId, messageProcessOptions.FilterChannelIDs) {
+		channel, appErr := p.API.GetChannel(post.ChannelId)
+
+		if appErr != nil {
+			p.API.LogError("Unable to get channel", "err", appErr)
```
Comment (hanzei) [inline plugin/helpers_bots.go]: Please leave it to the caller to log the error. Otherwise we will end up with two log messages.

Comment (RajatVaryani) [reply plugin/helpers_bots.go]: We have done the same thing in `EnsureBot`. i.e here https://github.com/mattermost/mattermost-server/blob/e5ba0a0a1808660d4885ef04888da437c2a5b6a0/plugin/helpers_bots.go#L55
Do you want me to fix it here as well?

File: plugin/helpers_bots.go (hunk 1 above)
Comment (RajatVaryani) [reply plugin/helpers_bots.go]: Even if we remove the `IsBotDMChannel` check, if I don't pass the `AllowSelf()` the `AllowSelf` variable remains false and the check goes to https://github.com/RajatVaryani/mattermost-server/blob/ac616eaeab10723c0f59b8bc3cea8ea840a435c6/plugin/helpers_bots.go#L123 Here the value of botUserId is needed.

File: plugin/helpers_bots.go (hunk 8 above)
Comment (hanzei) [reply plugin/helpers_bots.go]: The different in `EnsureBot` is this error is not returned. The rule is: either log the error or return it. The later is preferred.

Comment (RajatVaryani) [reply plugin/helpers_bots.go]: Okay. Thanks.

File: plugin/helpers_bots.go (hunk 1 above)
Comment (hanzei) [reply plugin/helpers_bots.go]: Hm, I'm wondering if `AllowSelf` is needed at all. https://github.com/RajatVaryani/mattermost-server/blob/ac616eaeab10723c0f59b8bc3cea8ea840a435c6/plugin/helpers_bots.go#L155-L159 will filter out any bot posts anyway. Do you see value in this check? Is there a case where you want to skip bot posts, but not your own bot?

Comment (RajatVaryani) [reply plugin/helpers_bots.go]: @hanzei Logically it makes sense. I will remove the `AllowSelf`.

File: plugin/helpers_bots.go (hunk 2 above)
Comment (RajatVaryani) [reply plugin/helpers_bots.go]: I think we can use `messageProcessOptions.FilterChannelIDs`. What do you think? The only downside the user of the helper method will have to keep track of all the channelIDs.

File: plugin/helpers_bots.go (hunk 8 above)
Comment (RajatVaryani) [reply plugin/helpers_bots.go]: I have made the changes.

File: plugin/helpers_bots.go (hunk 1 above)
Comment (RajatVaryani) [reply plugin/helpers_bots.go]: Removed Allowself.

File: plugin/helpers_bots.go (hunk 2 above)
Comment (RajatVaryani) [reply plugin/helpers_bots.go]: Just the decision on this is pending. Want to know your opinion.

**Comment by RajatVaryani** [comment]
Closing this PR. As my fork of the repo got deleted accidentally, Will raise another one.

**Comment by RajatVaryani** [comment]
Closing this as my fork got deleted accidentally. :(

**Pull request closed without merging**